src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -702,21 +702,29 @@ public void ExecuteCodeBlockActions<TLanguageKindEnum>(
         {
             ExecuteBlockActionsCore<CodeBlockStartAnalyzerAction<TLanguageKindEnum>, CodeBlockAnalyzerAction, SyntaxNodeAnalyzerAction<TLanguageKindEnum>, SyntaxNode, TLanguageKindEnum>(
                 codeBlockStartActions, codeBlockActions, codeBlockEndActions, analyzer,
-                declaredNode, declaredSymbol, executableCodeBlocks, (codeBlocks) => codeBlocks.SelectMany(
-                    cb =>
-                    {
-                        var filter = semanticModel.GetSyntaxNodesToAnalyzeFilter(cb, declaredSymbol);
-
-                        if (filter is object)
-                        {
-                            return cb.DescendantNodesAndSelf(descendIntoChildren: filter).Where(filter);
-                        }
-                        else
-                        {
-                            return cb.DescendantNodesAndSelf();
-                        }
-                    }),
+                declaredNode, declaredSymbol, executableCodeBlocks, addNodesToAnalyze,
                 semanticModel, getKind, filterSpan, isGeneratedCode, cancellationToken);
+
+            return;
+
+            void addNodesToAnalyze(ImmutableArray<SyntaxNode> codeBlocks, SpannableArrayBuilder<SyntaxNode> nodesToAnalyze)
+            {
+                foreach (var cb in codeBlocks)
+                {
+                    var filter = semanticModel.GetSyntaxNodesToAnalyzeFilter(cb, declaredSymbol);
+
+                    if (filter is object)
+                    {
+                        foreach (var node in cb.DescendantNodesAndSelf(descendIntoChildren: filter).Where(filter))
```

Comment:
considering calling filter directively on the node, to avoid linqisms.

Reply:
Sure, if you think that's more clear, I'll go ahead and switch to it. I don't think it will reduce allocations as DescendantNodesAndSelf already returns an IEnumerable, but it seems generally good to avoid linq in this code.